File: problem-archive/DP/L746_Min_Cost_Climbing_Stairs.cpp
/**
 * Leetcode 746) Min cost climbing stairs
 * @see https://leetcode.com/problems/min-cost-climbing-stairs/
 */
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    int minCostClimbingStairs(vector<int>& cost) {
        int size = cost.size();
        vector<int> cache = cost;

        for(int ri = size - 3; ri >= 0; ri--) {
            cache[ri] = cost[ri] + min(cache[ri + 1], cache[ri + 2]);
        }

        return min(cache[0], cache[1]);
    }
};

struct Testing {
    vector<int> input;
    int output;
    static const int TCASE_SIZE = 2;
};

int main() {
    Testing t[Testing::TCASE_SIZE] = {
        Testing{{10,15,20}, 15},
        Testing{{1,100,1,1,1,100,1,1,100,1}, 6},
    };

    Solution s;
    for (auto& tcase : t) {
        auto res = s.minCostClimbingStairs(tcase.input);
        if (res != tcase.output) {
            cout << "FAIL\n";
            cout << "Want: " << tcase.output << endl;
            cout << "Got: " << res << endl;
            return 1;
        }
    }

    cout << "ok\n";
    return 0;
}

File: problem-archive/DP/L746_Min_Cost_Climbing_Stairs.java
package com.java.app;

class Solution {
    public int minCostClimbingStairs(int[] cost) {
        int[] cache = new int[cost.length + 1];
        for (int i = 2; i < cache.length; i++) {
            cache[i] = Math.min(cost[i - 1] + cache[i - 1], cost[i - 2] + cache[i - 2]);
        }
        return cache[cache.length - 1];
    }
}

class Testing {
    public int[] cost;
    private int output;

    public Testing(int[] cost, int output) {
        this.cost = cost;
        this.output = output;
    }

    public boolean test(int output) {
        return this.output == output;
    }
}

public class L746_Min_Cost_Climbing_Stairs {
    public static void main(String[] args) {
        int[][] costs = {
            {10,15,20},
            {1,100,1,1,1,100,1,1,100,1},
        };

        Testing[] tcases = {
            new Testing(costs[0], 15),
            new Testing(costs[1], 6),
        };

        Solution s = new Solution();

        for (Testing t : tcases) {
            if (!t.test(s.minCostClimbingStairs(t.cost))) {
                System.err.println("FAIL");
                return;
            }
        }

        System.out.println("ok");
    }
}
